### ISSUE TITLE
ghcjs 8.4 javascript function name expression FFI creates an extra parameter with zero for JSVal and JSString

### ISSUE THREAD

**Issue by louispan** [ghcjs/ghcjs#666]
In GHCJS8.4 every JString comparison is always true.

Using the following ghcjs 8.4 stack.yaml,

```
resolver: nightly-2018-04-21

# ghcjs 8.4 doesn't include ghcjs-base
extra-deps:
 - git: git@github.com:ghcjs/ghcjs-base.git
   commit: 4560541fbab5b77ee34d0348ec04bb1dda8e5db3

# compiler: ghcjs-0.2.0.9011009_ghc-8.4.1
compiler-check: match-exact

setup-info:
  ghcjs:
    source:
      ghcjs-0.2.0.9011009_ghc-8.4.1:
           url: https://github.com/alexanderkjeldaas/ghcjs/releases/download/fghcjs-8.4.1/ghcjs.tar.gz
```

Code generated by javascript FFI creates an extra parameter with an extra zero for every string

Eg. Usage of [JSString Ord instance](https://github.com/ghcjs/ghcjs-base/blob/4560541fbab5b77ee34d0348ec04bb1dda8e5db3/Data/JSString.hs#L224):

```
instance Ord JSString where
  compare x y = compareStrings x y

compareStrings :: JSString -> JSString -> Ordering
compareStrings x y = tagToEnum# (js_compare x y +# 1#)
{-# INLINE compareStrings #-}

foreign import javascript unsafe
  "h$jsstringCompare" js_compare :: JSString -> JSString -> Int#
```

Get generated into the following (h$jsstringCompare has *4* parameters instead of *2*, the extra parameters passed as zero)

```
function h$$xh()
{
  var a = h$r1;
  var b = h$stack[(h$sp - 1)];
  h$sp -= 2;
  var c = h$jsstringCompare(b, 0, a, 0);
  var d = c;
  if((d === 1))
  {
    return h$e(h$baseZCGHCziShowzizdfShowBool2);
  }
  else
  {
    return h$e(h$baseZCGHCziShowzizdfShowBool4);
  };
};
```

Since [h$jsstringCompare](https://github.com/ghcjs/ghcjs-base/blob/4560541fbab5b77ee34d0348ec04bb1dda8e5db3/jsbits/jsstring.js#L1015) is implemented as follows, only the first two parameters are used, with the 2nd parameter as zero integer. The string comparison with zero integer will is always false, meaning h$jsstringCompare always return 0 from the function, which gets converted to EQ.

```
function h$jsstringCompare(x, y) {
    TRACE_JSSTRING("compare: '" + x + "' '" + y + "'");
    return (x<y)?-1:((x>y)?1:0);
}
```

This bug does not appear if I use the standard ghcjs 8.0.1 stack setup

```
resolver: lts-7.19

compiler: ghcjs-0.2.1.9007019_ghc-8.0.1
compiler-check: match-exact

setup-info:
  ghcjs:
    source:
      ghcjs-0.2.1.9007019_ghc-8.0.1:
           url: http://ghcjs.tolysz.org/ghc-8.0-2017-02-05-lts-7.19-9007019.tar.gz
           sha1: d2cfc25f9cda32a25a87d9af68891b2186ee52f9
```

Eg. GHCJS 8.0.1 generated code looks like this

```
function h$$pY()
{
  var a = h$r1;
  var b = h$stack[(h$sp - 1)];
  h$sp -= 2;
  var c = h$jsstringCompare(b, a.d1);
  var d = c;
  var e = h$tagToEnum(((d + 1) | 0));
  if((e.f.a === 3))
  {
    return h$e(h$baseZCGHCziShowzishows19);
  }
  else
  {
    return h$e(h$baseZCGHCziShowzishows20);
  };
};
```

@alexanderkjeldaas, since I'm using your ghcjs 8.4 setup, have you also noticed that all JSString comparisons are always EQ?

**Comment by louispan** [comment]
Could this be a problem with ghcjs-base instead? According to https://github.com/ghcjs/ghcjs/issues/602 "ghcjs-base is
now a normal package", which I take to mean that ghcjs 8.4 doesn't pre-package ghcjs-base.

I'm using ghcjs-base from
```
extra-deps:
 - git: git@github.com:ghcjs/ghcjs-base.git
   commit: 4560541fbab5b77ee34d0348ec04bb1dda8e5db3
```

Is there a fixed version for ghcjs 8.4 ?

edit: See below, it also happens with JSVal, so the bug has nothing to do with ghcjs-base.

**Comment by alexanderkjeldaas** [comment]
@louispan I'll have to check.. we're probably using the same setup you use, but I'm not 100% certain.

**Comment by louispan** [comment]
On closer inspection, the problem happens with JSVal as well, and with any function name expression type FFI.

eg. the following exhibits the bug
```
foreign import javascript unsafe
  "h$blah" js_blah :: JSVal -> JSVal
```

The following is ok
```
foreign import javascript unsafe
  "$r = h$blah($1);" js_blah :: JSVal -> JSVal
```

Is the FFI parser incorrectly thinking JSVal and JSString is a size-2 type?

**Comment by luite** [comment]
oh this is quite bad, it must be because of the calling convention change for `ByteArray#`, which now matches `Addr#`.

Unfortunately a `ByteArray#` primitive type is used internally for `JSVal` (and `JSString` is just a newtype wrapper of that).

I'm not sure yet what the most reliable way to fix this would be, but I think first the calling convention change should be restricted to `ccall`, so `foreign import javascript` shouldn't be affected at all.

**Comment by louispan** [comment]
I'm glad I've given enough details for you to identify the bug. I wasn't sure because [`ByteArray#`](https://github.com/ghcjs/ghcjs/blob/3959a9321a2d3e2ad4b8d4c9cc436fcfece99237/src/Gen2/ClosureInfo.hs#L156) varSize is [1](https://github.com/ghcjs/ghcjs/blob/3959a9321a2d3e2ad4b8d4c9cc436fcfece99237/src/Gen2/ClosureInfo.hs#L84)

**Comment by sheyll** [comment]
May I respectfully ask on the current status of this? Is it an upstream (GHC) problem, is there some kind of discussion happening?
Also, more generally, what is the general status/roadmap of the GHCJS 8.2, 8.4 (and upcoming 8.6)?

Is there something that I could help with?

**Comment by ryantrinkle** [comment]
I have a patch that seems to fix this, here: https://github.com/reflex-frp/reflex-platform/blob/4e669e6a9ab2b65e53da71eeace698d896aea3dd/haskell-overlays/ghcjs-base.patch

**Comment by cocreature** [comment]
@ryantrinkle That looks more like a workaround for `ghcjs-base` (but not for any FFI code using `JSString` that might be in other code bases) rather than a fix. I don’t think the solution here should be that for any FFI call that uses `JSString` you now need to add an additional (unused?) argument.

**Comment by louispan** [comment]
Also the bug is not limited to JSString. It also happens with JSVal and probably all newtype wrappers around JSVal

**Comment by dmjio** [comment]
yikes, any updates on this ?

**Comment by dalaing** [comment]
I'm about to test this out:

```
diff --git a/src/Gen2/Generator.hs b/src/Gen2/Generator.hs
index e9bf48a..08c3719 100644
--- a/src/Gen2/Generator.hs
+++ b/src/Gen2/Generator.hs
@@ -1690,7 +1690,7 @@ allocateStaticList _ _ = panic "allocateStaticList: unexpected literal in list"
 genFFIArg :: Bool -> StgArg -> G (JStat, [JExpr])
 genFFIArg isJavaScriptCc (StgLitArg l) = (mempty,) <$> genLit l
 genFFIArg isJavaScriptCc a@(StgVarArg i)
-    | tycon == byteArrayPrimTyCon || tycon == mutableByteArrayPrimTyCon = do
+    | not isJavaScriptCc && (tycon == byteArrayPrimTyCon || tycon == mutableByteArrayPrimTyCon) = do
         (\x -> (mempty,[x,jint 0])) <$> jsId i
     | isVoid r                  = return (mempty, [])
 --    | Just x <- marshalFFIArg a = x
```

but that restricts the calling convention to "anything but JavaScript calling convention" rather than "only the ccall calling convention".

I'm wondering if I should replace the `isJavaScriptCC :: Bool` parameter with something of type `CCallConv` in the generator to deal with this (and any future potentiall calling changes that might depend on the calling convention)? I might prepare branches that do both in the meantime and wait for @luite (or someone else who is across this kind of thing) to weigh in.

**Comment by dalaing** [comment]
Hmm, I'm getting errors from `assignAllCh` now, where the expression context has 1 argument for something and `genIds` is creating 2 arguments based on the type.  I guess we _could_ go down the path of having the sizes of types be based on calling conventions, but that might also be a bit on the crazy side.

**Comment by dalaing** [comment]
Ah, whoops, I was just running into #665 :)

So with [this change](https://github.com/dalaing/ghcjs/commit/0ffddbb74101b5bcf26563d6834cfaafc540cd51) and this program:

```haskell
{-# LANGUAGE OverloadedStrings #-}
module Main where

import Data.JSString

main :: IO ()
main = do
  let x = ("a" :: JSString) < ("b" :: JSString)
  print x
```

I get the following output:

```
function h$mainZCMainzimain2_e()
{
  var a = h$r1.d1;
  var b = h$r1.d2;
  h$bh();
  var c = "a";
  var d = c;
  var e = "b";
  var f = h$jsstringCompare(d, e);
  var g = f;
  var h = h$tagToEnum(((g + 1) | 0));
  if((h.f.a === 1))
  {
    return h$e(a);
  }
  else
  {
    return h$e(b);
  };
};
```

Woot?

**Comment by luite** [comment]
I added an alternative fix to the 8.4 branch, and I think the main difference is that my fix uses the new calling convention for everything that's not `JavaScriptCc`. @dalaing 's patch uses the new calling convention for `ccall`

Which one is to be preferred?

**Comment by dalaing** [comment]
I can't remember what I tried when I wrote this, but I know a couple of attempts didn't work / ran into other walls.  

I had [this](https://github.com/dalaing/ghcjs/commit/4b99037d9f77930a30dbf678653769719ae61a62) at one point and backed away from it in favour of [this](https://github.com/dalaing/ghcjs/commit/0ffddbb74101b5bcf26563d6834cfaafc540cd51), but I can't remember why I switched between the two approaches.  I think I was testing it by bootstrapping ghcjs, and I might have also been building `reflex` at the time as well.

I don't have any real preference, as long as the fix works :)
